py/modgc.c

```diff
@@ -32,7 +32,9 @@
 
 // collect(): run a garbage collection
 STATIC mp_obj_t py_gc_collect(void) {
+    GC_ENTER();
     gc_collect();
+    GC_EXIT();
```

Comment:
It's probably safer to put these GC_ENTER/GC_EXIT pairs in the `gc_collect()` implementation, because `gc_collect()` may be called by C code from another location, see eg https://github.com/micropython/micropython/blob/master/lib/utils/pyexec.c#L141

Reply:
How do we best accomplish that? gc_collect() is used inside gc_alloc(), and we cannot release GC mutex without a segfaulty race.

Reply:
maybe we rename the funciton to _gc_collect() in all the ports, and create a macro gc_collect(), which has GC_ENTER/EXIT wrapped around it?